assets/js/wfparameters/Refine.js

```diff
@@ -0,0 +1,383 @@
+            histogramNumRows: 0,
+            showWarning: false,
+            edits: JSON.parse(props.existingEdits.length > 0 ? props.existingEdits : '[]'),
+        }
+
+        this.handleValueChange = this.handleValueChange.bind(this);
+        this.handleSelectionChange = this.handleSelectionChange.bind(this);
+
+        //console.log(this.state.edits);
+    }
+
+    componentDidMount() {
+        this.loadHistogram(this.props.selectedColumn, this.state);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        // Handles revision changes and column changes
+
+        //console.log(nextProps);
+        var nextColumn = nextProps.selectedColumn;
+        var nextRevision = nextProps.revision;
+        if(nextRevision != this.props.revision) {
+            //console.log(nextRevision, this.props.revision);
+            //console.log('Revision bumped.');
+            if(nextColumn != this.props.selectedColumn) {
+                // If the column changes, check if this is a cancel; if not, clear the edits
+                // The empty edits will be saved to the server on the next edit
+                this.setState({
+                    histogramLoaded: false,
+                    histogramData: [],
+                    histogramNumRows: 0,
+                    showWarning: nextRevision > this.props.revision,
+                    edits: (nextRevision > this.props.revision) ? [] : JSON.parse(nextProps.existingEdits),
+                }, () => {this.loadHistogram(nextColumn, this.state)});
+            } else {
+                /*
+                this.setState({
+                    histogramLoaded: false,
+                    histogramData: [],
+                    histogramNumRows: 0,
+                    showWarning: false,
+                    edits: JSON.parse(nextProps.existingEdits.length > 0 ? nextProps.existingEdits : '[]'),
+                }, () => {
+                    //console.log(this.state.edits);
+                    this.loadHistogram(nextColumn);
+                });
+                */
+                var nextState = Object.assign({}, this.state);
+                nextState.edits = JSON.parse(nextProps.existingEdits.length > 0 ? nextProps.existingEdits : '[]');
+                console.log(nextState.edits);
+                this.loadHistogram(nextColumn, nextState);
+            }
+        }
+    }
+
+    loadHistogram(targetCol, baseState, clearEdits=false) {
+        // Loads a histogram from the server and sets the state with the result
+
+        // clearEdits controls whether we clear the edit on the server immediately upon load
+        // This is unused for now.
+        api.histogram(this.props.wfModuleId, targetCol)
+            .then(histogram => {
+                //console.log(histogram);
+                var nextState = Object.assign({}, baseState);
+                var editedHistogram = histogram.rows.map(function(entry) {
+                    var newEntry = Object.assign({}, entry);
```

Comment:
Check out the update() package for this sort of thing.  https://www.npmjs.com/package/immutability-helper